enhancements/console/catalog-categories.md

```diff
@@ -0,0 +1,153 @@
+---
+title: catalog-categories
+authors:
+  - "@jerolimov"
+reviewers:
+  - TBD
+  - "@spadgett"
+  - "@cvogt"
+approvers:
+  - TBD
+creation-date: 2020-10-22
+last-updated: 2020-10-22
+status: implementable
+---
+
+# Customize Developer Catalog Categories
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+The OpenShift Console provides a "Developer Catalog" which enables the user to create apps and services based on Operator backed services, helm charts, templates, etc.
+
+Cluster admins can provide additional catalog items by installing operators or adding helm repositories. But the filter categories (All Items, Languages, Databases, Middleware, ...) are hard coded. This enhancement enable cluster admin to customize these category filters.
+
+## Motivation
+
+Customers who provide their own catalog items would like to customize the category filters such that they better represent their catalog items and make it easier for developers to navigate the catalog.
+
+### Goals
+
+- Extract the Developer Catalog categories into a CRD to allow cluster admins to customize them.
+
+### Non-Goals
+
+- Hiding categories based on the available catalog items.
+- Introducing a UI to configure this categories without a yaml editor.
+
+## Proposal
+
+### Extending the Console CRD
+
+Extend the existing `operator.openshift.io/v1` / `Console` CRD. It provides already a `spec.customization` area where we can add a list of available categories.
+
+Each category could have a list of subcategories which enables the Developer Catalog to build a 2 level tree based on this data:
+
+```
+├── Languages
+│   ├── Java
+│   ├── JavaScript
+│   └── ...
+├── Databases
+├── Middleware
+│   ├── Integration
+│   ├── Process Automation
+│   └── ...
+└── ...
+```
+
+Each Category and Subcategories matches following schema:
+
+* `id: string`, identifier used in the URL to enable deeplinking in console
+* `label: string`, category display label
+* `disabled: boolean`, flag to hide this category from the sidebar so that the admin is not required to remove a complete section to hide an unused category
+* `tags: string[]`, the list of item tags this category will match
+
+Categories on the first level contains also a list of subcategories:
+
+* `subcategories:` a list of Categories
+
+If a category has subcategories the dev console will show all tags from a selected category and all subcategories.
+
+So the final yaml should look like this one:
+
+```yaml
+apiVersion: operator.openshift.io/v1
+kind: Console
+metadata:
+  annotations:
+    release.openshift.io/create-only: 'true'
+  name: cluster
+  ...
+spec:
+  customization:
+    brand: online
+    devcatalog:
+      categories:
```

Comment:
Does this completely replace the categories or just add new categories?

Reply:
Updated the document that we use the hard coded version OR the customization version. Ok?